### PR TITLE
Update to the 5.0 APIs, and now pass in the quarantine environment variables

### PR DESCRIPTION
**Pull request by itay**
I've updated the code to properly use the new 5.0 hook APIs, as well as pass in the environment variables. This will make it work with the newer versions of Git.

I'm not sure how to properly update the pom.xml here, so I will leave that to @seletskiy or someone else who can suggest. The plugin may need to be marked as 5.0+ only now as well, though I'm not sure.

@bturner if you're able to take a look and give feedback, that would be very welcome.

### CODE REVIEW

**Comment by seletskiy** [comment]
@itay: Please, make PR to https://github.com/seletskiy/atlassian-external-hooks, I do not have access to that repo anymore.

File: src/main/java/com/ngs/stash/externalhooks/hook/ExternalAsyncPostReceiveHook.java [hunk 1]
```diff
@@ -36,15 +33,12 @@ public ExternalAsyncPostReceiveHook(
...
-        Collection<RefChange> refChanges
-    ) {
+	@Override
+	public void postUpdate(PostRepositoryHookContext context, RepositoryHookRequest request) {
         ExternalPreReceiveHook impl = new ExternalPreReceiveHook(
-            this.authCtx, this.permissions, this.repoService, this.properties);
-        impl.onReceive(context, refChanges, null);
-    }
+                this.authCtx, this.permissions, this.repoService, this.properties);
+            impl.preUpdateImpl(context, request);
```
Comment (bturner) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalAsyncPostReceiveHook.java]: This indenting seems off /nit

File: src/main/java/com/ngs/stash/externalhooks/hook/ExternalMergeCheckHook.java [hunk 2]
```diff
@@ -59,34 +55,28 @@ public ExternalMergeCheckHook(
...
-    @Override
-    public void check(
-        RepositoryMergeRequestCheckContext context
-    ) {
-        PullRequest pr = context.getMergeRequest().getPullRequest();
+
+
+	@Override
+	public RepositoryHookResult preUpdate(PreRepositoryHookContext context, PullRequestMergeHookRequest request) {
+		// TODO Auto-generated method stub
```
Comment (bturner) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalMergeCheckHook.java]: This, and the blank line after it, should probably be removed?

File: src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java [hunk 3]
```diff
@@ -99,6 +93,10 @@ public ProcessBuilder createProcessBuilder(
             log.error("Can't get user email address. getEmailAddress() call returns null");
         }
         env.put("STASH_REPO_NAME", repo.getName());
+        
+        if (request.getScmHookDetails().isPresent()) {
```
Comment (bturner) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java]: I'd write this as:

```java
request.getScmHookDetails().ifPresent(details -> env.putAll(details.getEnvironment()));
```

File: src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java [hunk 4]
```diff
@@ -166,33 +164,37 @@ public int runExternalHooks(
...
-                        trimmed = true;
-                    }
-                    continue;
+        int data;
+        int count = 0;
+        StringBuilder builder = new StringBuilder();
+        while ((data = input.read()) >= 0) {
+            if (count >= 65000) {
+                if (!trimmed) {
+                	builder.
```
Comment (bturner) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java]:
```java
builder.append("\n")
        .append("Hook response exceeds 65K length limit.\n")
        .append("Further output will be trimmed\n");
trimmed = true;
```

Maybe?

File: src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java [hunk 5]
```diff
@@ -166,33 +164,37 @@ public int runExternalHooks(
...
-            }
+            count += charToWrite.getBytes("utf-8").length;
 
+            builder.append(charToWrite);
+        }
+        
+        int result = process.waitFor();
+        
+        if (result == 0) {
+        	return RepositoryHookResult.accepted();
```
Comment (bturner) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java]: Indenting /nit

File: src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java [hunk 6]
```diff
@@ -166,33 +164,37 @@ public int runExternalHooks(
...
+        } else {
```
Comment (bturner) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java]: This `else` is unnecessary. The `if` above it `return`s, so you could just write:

```java
if (result == 0) {
    return RepositoryHookResult.accepted();
}
return RepositoryHookResult.rejected(summaryMessage, builder.toString());
```

File: src/main/java/com/ngs/stash/externalhooks/hook/ExternalMergeCheckHook.java [hunk 7]
```diff
@@ -137,44 +127,34 @@ public void check(
         String summaryMsg = "Merge request failed";
 
         try {
-            int Result = runExternalHooks(pb, refChanges, hookResponse);
-            if (Result != 0) {
-                String rawDetailedMsg = errWriter.toString();
-                String prePrefix = "<pre style=\"overflow: auto; white-space: nowrap;\">";
-                String preSuffix = "</pre>";
-                String detailedMsg = prePrefix + rawDetailedMsg.replaceAll("(\r\n|\n)", "<br/>").replaceAll(" ", " ") + preSuffix;
```
Comment (bturner) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalMergeCheckHook.java]: Does dropping this `detailedMsg` formatting affect how this displays in the UI when the check fails? Is that desirable? (Or perhaps the new hook SPI doesn't allow for the formatting to work correctly, anymore? I could believe that.)

Comment (bturner) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalMergeCheckHook.java]: (Nevermind; I see from #53 that dropping this is a good thing to do. It sounds like @seletskiy has already made the matching change to his version of the plugin.)

File: src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java (hunk 3 above)
Comment (itay) [inline src/main/java/com/ngs/stash/externalhooks/hook/ExternalPreReceiveHook.java]: Thanks. Fancy Java is new to me :)

**Comment by itay** [comment]
@bturner thanks for the feedback.

@seletskiy I will correct these and open a new PR against your repo!

**Comment by fncmedia** [comment]
@itay  I sounds like your fix could solve our current problems. Have you already a timing for finishing your pr?

**Comment by seletskiy** [comment]
@itay: ping.

**Comment by joerg** [comment]
Hi,
I just stumbled upon this because our pre-recieve hooks are not working any more. Seems the missing GIT_QUARANTINE_PATH is the cause. Any chance to get this fixed?

**Comment by chroto** [comment]
We also are struggling with this and would like to see it merged.  Anything I can help with?

**Comment by chroto** [comment]
@itay @seletskiy would it be in bad form to open this PR on https://github.com/seletskiy/atlassian-external-hooks myself ?

**Comment by itay** [comment]
Please go ahead :) No reason to block this on my unavailability!

**Comment by DrVanScott** [comment]
I added this PR on seletskiy's some days ago. Hope i did it the right way... https://github.com/seletskiy/atlassian-external-hooks/pull/3

**Comment by seletskiy** [comment]
Yep, everything is fine, I just managed to obtain ownership over repository
back.
I'm going to release new version soon, so stay in touch :)

**Comment by fncmedia** [comment]
@seletskiy   can you make an estimate when you would update your plugin?

**Pull request merged**